자동차 경주 4단계

bsgreentea opened the pull request:
작업 내용
- Car 클래스에 name 프로퍼티를 추가했습니다.
- 들여쓰기 깊이가 깊어지지 않도록 다듬었습니다.
- 3단계 리뷰를 반영했습니다.

질문
RacingGame 클래스의 startGame 메서드에 파라미터를 추가했는데요, 테스트를 위해 입력값을 인자로 받을 수 있도록 한 것이 목적입니다.
기능적으로는 굳이 나눌 필요가 없을 것 같은데 억지로 나눈 감이 있네요. 이 경우 입력부와 수행부(?)를 나누는 좋은 방법이 있을지 궁금합니다.

bsgreentea committed:
자동차 경주 2단계 - 문자열 계산기 (#1332)

* feat: Calculator 클래스 선언 및 사칙연산을 위한 메서드 구성

* test: Calculator 클래스 사칙연산 메서드에 대한 테스트코드 작성

* feat: 입력값과 연산자 타입 확인할 수 있는 검증 코드 추가

* test: 입력 내용이 빈칸인지와 연산자가 올바른지에 대한 테스트 코드 추가

* feat: 입력 식 계산 메서드 구현

* test: 식 계산에 대한 테스트 코드 추가

* refactor : 연산자 체크 로직과 테스트 코드 수정해라

- 연산자인지 확인하는 방법을 '한 글자이고 숫자일 때' 하는 것으로 변경함.
- OperationType > from으로 타입을 가져올 때, 정의되지 않은 값이라면 IllegalArgumentException 던지도록 수정
- 연산자 검증하는 테스트를 위해 OperatorTypeTest 생성
- 기존의 calculate 테스트는 `식 계산` 메서드가 역할을 하고 있기 때문에 제거함

* refactor : 리뷰 반영해라

- 연산 수행을 OperatorType에서 하도록 수정
- 각 OperatorType은 OperatorExecutor의 execute를 구현하도록 함
- Calculator를 object로 변경
- Calculator의 calculator 메서드를 process로 리네이밍함. 클래스와 중복되는 이름같아 불필요하다고 생각했음.
- 테스트 코드에 @ParameterizedTest 적용

---------

Co-authored-by: bsgreentea <>

bsgreentea committed:
자동차 경주 2단계 - 문자열 계산기 (#1332)

* feat: Calculator 클래스 선언 및 사칙연산을 위한 메서드 구성

* test: Calculator 클래스 사칙연산 메서드에 대한 테스트코드 작성

* feat: 입력값과 연산자 타입 확인할 수 있는 검증 코드 추가

* test: 입력 내용이 빈칸인지와 연산자가 올바른지에 대한 테스트 코드 추가

* feat: 입력 식 계산 메서드 구현

* test: 식 계산에 대한 테스트 코드 추가

* refactor : 연산자 체크 로직과 테스트 코드 수정해라

- 연산자인지 확인하는 방법을 '한 글자이고 숫자일 때' 하는 것으로 변경함.
- OperationType > from으로 타입을 가져올 때, 정의되지 않은 값이라면 IllegalArgumentException 던지도록 수정
- 연산자 검증하는 테스트를 위해 OperatorTypeTest 생성
- 기존의 calculate 테스트는 `식 계산` 메서드가 역할을 하고 있기 때문에 제거함

* refactor : 리뷰 반영해라

- 연산 수행을 OperatorType에서 하도록 수정
- 각 OperatorType은 OperatorExecutor의 execute를 구현하도록 함
- Calculator를 object로 변경
- Calculator의 calculator 메서드를 process로 리네이밍함. 클래스와 중복되는 이름같아 불필요하다고 생각했음.
- 테스트 코드에 @ParameterizedTest 적용

---------

Co-authored-by: bsgreentea <>

bsgreentea committed:
feat: 자동차 경주 입력값 받을 수 있게 해라

입력값 받을 수 있도록 구성

bsgreentea committed:
feat: 자동차 경주 구현해라

- Car 클래스에서 차의 주행 상태를 관리하도록 구성
- 랜덤 수를 생성하는 메서드는 RandomUtils로 분리
- ResultView에서 결과 출력하도록 구현

bsgreentea committed:
rafctor: 2단계 리뷰 반영해라

- OperatorExecutor를 상속하는 대신 SAM으로 사용하도록 변경. 코드가 좀 더 간결해졌다!
- 나누기 0은 기존에도 익셉션을 던지므로 그냥 계산하도록 수정함. 에러 나면 종료하도록.
- 클래스명, 메서드명, 파라미터 등 리네이밍
- 연산자별 연산 테스트 시 @MethodSource 적용

bsgreentea committed:
rafctor: 2단계 리뷰 반영해라

- OperatorExecutor를 상속하는 대신 SAM으로 사용하도록 변경. 코드가 좀 더 간결해졌다!
- 나누기 0은 기존에도 익셉션을 던지므로 그냥 계산하도록 수정함. 에러 나면 종료하도록.
- 클래스명, 메서드명, 파라미터 등 리네이밍
- 연산자별 연산 테스트 시 @MethodSource 적용

bsgreentea committed:
Merge remote-tracking branch 'upstream/bsgreentea' into step3

# Conflicts:
#	src/main/kotlin/week1/calculator/Calculator.kt
#	src/main/kotlin/week1/calculator/OperatorExecutor.kt
#	src/test/kotlin/week1/calculator/CalculatorTest.kt

bsgreentea committed:
Update README.md

자동차 경주 4단계 구현해야할 기능 체크리스트 작성

bsgreentea committed:
Update README.md

개행 정리

bsgreentea committed:
feat: Car에 이름 프로퍼티 추가 및 이름 관련 입력 영역을 수정한다.

- 자동차 이름을 입력받을 수 있도록 입력 로직 수정
- Car에 name 프로퍼티 추가
- README.md 업데이트

bsgreentea committed:
update: Update README.md

- 우승자 판단 로직이 추가되어 리드미 수정

bsgreentea committed:
feat: 결과 출력 시 자동차 이름을 노출하고 마지막에 우승자를 나타낸다.

bsgreentea committed:
refactor: 3단계 리뷰를 반영한다.

- 입력 시 널체크 메서드를 requireNotNull로 변경하고 에러 메시지를 추가한다.
- RacingGame이 게임 횟수와 차 정보에 대한 프로퍼티를 갖지 않도록 한다. 레이스에 집중한다.
- RandomUtils는 제거. 클래스까지는 필요 없을듯.
- 매직 넘버는 const val로 관리.
- 횟수 반복을 위한 forEach문은 repeat으로 변경한다.
- 메서드 내 들여쓰기 깊이가 2를 넘지 않도록 한다.

bsgreentea committed:
test: Car 클래스의 생성자와 메서드를 테스트한다.

bsgreentea committed:
test: 입력값과 경주에 대해 테스트한다.

bsgreentea committed:
lint: 포맷팅한다.

bsgreentea committed:
Merge remote-tracking branch 'upstream/bsgreentea' into step4

# Conflicts:
#	src/main/kotlin/week1/calculator/Calculator.kt
#	src/main/kotlin/week1/calculator/Operator.kt
#	src/main/kotlin/week1/racingcar/Car.kt
#	src/main/kotlin/week1/racingcar/InputResult.kt
#	src/main/kotlin/week1/racingcar/InputView.kt
#	src/main/kotlin/week1/racingcar/RacingGame.kt
#	src/main/kotlin/week1/racingcar/ResultView.kt
#	src/test/kotlin/week1/calculator/OperatorTest.kt

bsgreentea committed:
refactor: 실행 순서 수정해라

- 문구 출력 순서 변경

Rok93 commented on README.md:
```diff
@@ -1 +1,9 @@
+
+## 자동차 경주 4단계
+
+구현할 기능들
```
> RacingGame 클래스의 startGame 메서드에 파라미터를 추가했는데요, 테스트를 위해 입력값을 인자로 받을 수 있도록 한 것이 목적입니다. 기능적으로는 굳이 나눌 필요가 없을 것 같은데 억지로 나눈 감이 있네요. 이 경우 입력부와 수행부(?)를 나누는 좋은 방법이 있을지 궁금합니다.

먼저 위와 같은 질문을 주셨는데, 자동차 경주게임 5단계 미션 요구사항을 살짝 보시면 좋을 것 같네요. 😃 
(사실 다음 단계에서 다루게 될 이야기였는데, 벌써 그런 니즈를 느끼시다니.. 👍👍💯😉)

> MVC 패턴 기반으로 리팩토링해 view 패키지의 객체가 domain 패키지 객체에 의존할 수 있지만, domain 패키지의 객체는 view 패키지 객체에 의존하지 않도록 구현한다

자동차 경주 게임 미션을 `M`(Model), `V`(View), `C`(Controller)로 분리하게되면 질문주신 `입력부`(View)와 `수행부`(Model)로 자연스럽게 나눠질 것 같습니다. 😃 

MVC 패턴으로 분리하는 내용은 다음 단계에서 집중적으로 다룰테니 이번 단계에서는 도전해보고싶으시다면 진행해보셔도 좋을 것 같아요. 😉

혹시라도 MVC 패턴이 익숙하지 않다면 아래의 영상을 참고해보시면 좋을 것 같아요. 😉
[[10분 테코톡] 🧀 제리의 MVC 패턴](https://www.youtube.com/watch?v=ogaXW6KPc8I)

Rok93 commented on src/main/kotlin/week1/racingcar/InputView.kt:
```diff
@@ -1,15 +1,38 @@
 package week1.racingcar
 
 class InputView {
```
InputView, ResultView는 프로퍼티를 가지지 않아서 굳이 객체를 생성하지 않아도될 것 같아요. 🤔
이번 기회에 [object](https://kotlinlang.org/docs/object-declarations.html) 키워드의 특징을 알아보시고, 사용해보시면 좋을 것 같아요. 🤗

Rok93 commented on src/main/kotlin/week1/racingcar/InputView.kt:
```diff
@@ -1,15 +1,38 @@
+            "자동차 이름은 null을 허용하지 않습니다."
+        }
+        carNames.forEach {
+            require(it.isNotBlank()) {
+                "자동차 이름은 빈 칸을 허용하지 않습니다."
+            }
+            require(it.length <= 5) {
+                "자동차 이름은 5자를 초과해선 안됩니다."
+            }
```
`자동차의 이름`의 유효성 검증은 어디에서 수행하는 것이 좋을까요? 🤔

만약  요구사항이 바껴서 게임 시작시에 경주용 차량을 생성하는 것 외에도 추가로 자동차를 생성하는 절차가 생긴다면 어떻게 될까요? 
아마도 InputView 안에 새로운 메서드가 생기고, 지금과 똑같은 유효성 검증 로직이 추가될거에요. 👀 
시간이 지남에 따라 요구사항이 늘어나게되면 이렇게 똑같은 유효성 검증 로직은 계속해서 복붙이될거에요. 👀

그런데 어느날 요구사항이 바껴서 자동차 이름은 공백을 허용한다거나 혹은 8자까지는 허용한다와 같이 제약사항이 바뀌게된다면 변경해야할 위치가 엄청나게 많아질거에요. 🥲😇

그렇다면 이러한 자동차 객체의 이름에 대한 유효성 검증은 어디에서하는게 좋을까요? 🤔
그 답은 자동차 객체가 생성되는 시점에 하는 것이 좋을거에요.  
객체가 생성되는 시점에 유효성을 검증하도록 변경해보시면 좋을 것 같아요. 😃
또한 변경해보시고 어떤 장점들이 있는지 고민해보시면 좋을 것 같아요. 😉
고민해보시고 코멘트로 답변을 달아주시면 좋을 것 같습니다. 😆
(틀려도 괜찮으니 그냥 생각나는대로 달아주셔도 괜찮습니다)

bsgreentea replied:
한 지점에서 관리하는게 좋겠네요.
코드 수정해보니 확실히 관리 포인트가 줄어든 것 같네요!

Rok93 commented on src/main/kotlin/week1/racingcar/Car.kt:
```diff
@@ -1,10 +1,12 @@
+class Car(
+    val name: String,
+) {
     var mileage = 0
```
`0` 이라는 숫자가 어떤 값을 가지는지 직관적으로 이해하기가 어려운 것 같아요. 🥲
의미가 잘 드러나는 상수로 추춣보시면 어떨까요? 🤗

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
-            }
-            race()
+    fun getInputResult(): InputResult {
+        return InputView().let {
+            it.getValidateRaceInfo(it.startInput())
         }
```
작성하신 코드를 조금 풀어써보면 아래와 같아요

```kotlin
        val inputView = InputView()
        val input = inputView.startInput()
        inputView.getValidateRaceInfo(input)
```

`getValidateRaceInfo`와 `startInput` 메서드 둘 다  InputView의 메서드들인데, 따로 호출하고 호출한 값을 메서드 파라미터로 주입해줄 이유가 있을까요🤔 

그냥 하나로 합치면 되지 않을까요? 😃

bsgreentea replied:
getValidateRaceInfo 를 테스트한다고 억지로 빼놨는데요,
말씀해주신 것 처럼 굳이 나눌 필요가 없기도 하고 고민했던 부분인 테스트코드도 잘 통과한 것 확인했으니 합쳐도 상관 없겠군요!

Rok93 commented on src/main/kotlin/week1/racingcar/InputView.kt:
```diff
@@ -1,15 +1,38 @@
\ No newline at end of file
+
+    data class InputString(
+        val cars: String?,
+        val times: String?,
```
InputString은 DTO 클래스로 사용하신 것 같아요. 👀
조금 더 의미있는 네이밍으로 변경해보면 어떨까요?

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
 
-    private var numberOfCars = 0
-    private var timesToRace = 0
+class RacingGame {
```
자동차 경주 게임이라는 도메인에서는 `RacingGame` 객체가 중요한 도메인 모델이라고 생각해요. 👀
일단은 RacingGame을 객체답게 변경해보면 어떨까요? 😉
RacingGame이 레이스를 진행하기 위해서는 어떠한 객체들과 협력해야할까요? 
협력해야하는 객체들을 프로퍼티로 가지도록 변경해보고 주 생성자를 통해서 프로퍼티를 초기화할 수 있도록 변경해보면 어떨까요? 

(이렇게 변경하게되면 질문주셨던 구현부와 입력부를 분리하고자하는 니즈에도 한발 가까워질 것 같네요. 😃)

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
     }
 
-    private fun race() {
+    fun startGame(inputResult: InputResult) {
+        val cars = inputResult.cars.map { Car(it) }
+        val timesToRace = inputResult.timesToRace
+        race(cars, timesToRace)
+    }
```
사실상 `자동차 경주 게임` 객체를 생성하기위해서 셋팅하는 부분이라고 생각하는데, 
객체의 생성은 생성자에게 위임해주세요! 😃

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
+        }
+    }
+
+    private fun showResult(resultView: ResultView, cars: List<Car>) {
```
메서드의 호출 순서에 맞춰서 메서드를 위치시키면 조금 더 코드를 읽기가 편할 것 같아요. 😃
```kotlin
    private fun isPossibleToDrive() = Random.nextInt(COUNT_OF_DIGITS) >= MIN_NUM_TO_GO

    private fun showResult(resultView: ResultView, cars: List<Car>) {
        resultView.run {
            showResult(cars)
            showWinner(getWinners(cars))
        }
    }
```

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
+
+    companion object {
+        private const val MIN_NUM_TO_GO = 5
+        private const val COUNT_OF_DIGITS = 10
```
COUNT_OF_DIGITS가 무슨 의미인지 잘 모르겠어요. 🥲

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
+
+    private fun getWinners(cars: List<Car>): List<String> {
+        val farthestMileage = getFarthestMileage(cars)
+        return cars.filter { it.mileage == farthestMileage }.map { it.name }
```
가급적 하나의 하나의 `.`를 사용하도록 해보세요! 😃
가독성도 좋아지고 또 chain이 3개이상 연결되면 인텔리제이가 반환타입을 각 라인에 명시해줘서 코드를 파악하기 쉽게 해줍니다. 🤗

```suggestion
        return cars.filter { it.mileage == farthestMileage }
                .map { it.name }
```

<img width="559" alt="image" src="https://github.com/next-step/kotlin-racingcar/assets/47850258/06166c4f-e194-4d48-a78b-00526cf65053">

Rok93 replied:
객체의 프로퍼티를 외부로 꺼내서 직접 비교하는 것이 아니라 아래와 같이 객체에게 메시지를 보내보면 어떨까요? 🤗
```suggestion
        return cars.filter { it.isSameMileage(farthestMileage) }.map { it.name }
```

아래의 글을 참고해보시면 좋을 것 같아요. 😃
[getter를 사용하는 대신 객체에 메시지를 보내자](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/)

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
+        return cars.filter { it.mileage == farthestMileage }.map { it.name }
+    }
+
+    private fun getFarthestMileage(cars: List<Car>): Int {
```
이정도의 로직은 굳이 메서드로 쪼개지 않아도 충분히 가독성이 좋을 것 같아요. 🤔

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
+    }
+
+    private fun drive(cars: List<Car>) {
+        cars.forEach {
+            if (isPossibleToDrive()) it.go()
+        }
```
자동차가 전진 혹은 정지하는 것은 자동차의 역할인 것 같아요. 

조금 더 객체지향 코드를 작성해보도록 수정해보면 좋을 것 같은데 RacingGame 객체가 직접 전진할 조건을 계산하고 만족하면 자동차 객체를 전진하게 하는 것이 아니라 RacingGame 객체가 자동차 객체에게 `전진 혹은 정지해`라고 요청만하고 그 요청에 대한 수행은 자동차 객체가 스스로 판단해서 수행하도록 변경해볼 수 있을 것 같아요. 

```suggestion
        cars.forEach { it.moveOrNot }
```

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
-            }
-            resultView.showResult(cars)
+        repeat(timesToRace) {
+            drive(cars)
```
메서드를 분리를 조금 과하게 했다고 느껴지는 것 같아요. 🤔
저라면 지금처럼 메서드로 쪼개지않고 아래와 같이 바로 코드를 짰을 것 같아요. 👀

```suggestion
            cars.forEach {
                if (Random.nextInt(COUNT_OF_DIGITS) >= MIN_NUM_TO_GO) {
                    it.go()
                }
            }
```

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -1,36 +1,62 @@
-    RacingGame().start()
-}
\ No newline at end of file
+    RacingGame().run {
```
`let`, `run` scope function을 자주 사용하셨던데, 혹시 사용하시는 기준이 있으실까요? 

[Kotlin Scope function 올바르게 사용하기](https://hyeon9mak.github.io/using-kotlin-scope-function-correctly/) 글을 참고해보시면 좋을 것 같아요. 😉

bsgreentea replied:
어떤 값을 리턴해야하는 상황이 아니라면 run과 let 선택은 가독성을 기준으로 두고 있었습니다. 그런데 이건 제 기준이라 좋지 않을 수가 있겠네요ㅎㅎ
공유해주신 글 잘 읽어보겠습니다!

Rok93 reviewed (changes requested):
보성님 4단계 잘 구현해주셨습니다 👍👍 

몇몇 코멘트 남겨두었으니 확인해주셔서 반영해주시면 될 것 같습니다. 🙏
객체지향적인 코드로 변경해보기위해 추가한 코멘트들이 있는데, 변경하시면서 어려운 부분이 있으면 언제든지 DM 주세요. 🤗

bsgreentea committed:
refactor: 4차 리뷰를 반영한다.

- 자동차의 정보에 대한 값 검증을 Car 객체 초기화 시점에 하도록 수정
- 경주 관련 정보는 RacingGame에서 관리하도록 cars와 timesToRace 프로퍼티 추가
- 체이닝 메서드는 한 줄에 하나씩 나타내도록 수정
- 프로퍼티가 없는 클래스를 object로 변경
- 사용하지 않게 된 테스트코드 제거

Rok93 commented on src/test/kotlin/week1/racingcar/CarTest.kt:
```diff
@@ -0,0 +1,12 @@
+    @Test
+    fun `생성자 확인`() {
+        val car = Car("Thomas")
+        car.name shouldBe "Thomas"
```
테스트가 깨지는 것 같아요. 👀

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -2,35 +2,49 @@ package week1.racingcar
-                timesToRace = it.timesToRace
-            }
-            race()
+    private val cars: List<String>
```
RacingGame은 자동차의 이름들과 협력하는 것이 아니라, 자동차와 협력할 것 같아요. 👀
프로퍼티로 두면 어떨까요? 😃

또한 프로퍼티를 주 생성자를 통해서 초기화하도록 변경해보시면 RacingGame 객체에 대한 테스트 작성도 쉬워질 것 같습니다. 😉

```kotlin
class RacingGame(
    private val cars: List<Car>,
    private val timesToRace: Int,
) {
```

지금 현재는 RacingGame 클래스에 대한 테스트가 없는 것 같아요. 👀
다음 리뷰 요청 때는 꼭 테스트 코드도 보충해서 리뷰요청 부탁드려요! 🙏

Rok93 replied:
추가적으로 입력을 받는 것이  RacingGame 객체의 역할이 맞을지 고민해보시면 좋을 것 같아요.  😃
[해당 코멘트](https://github.com/next-step/kotlin-racingcar/pull/1469/files#r1380060911)에 대해서 한번 더 


추가적으로 [웹 MVC 각 컴포넌트 역할](https://tecoble.techcourse.co.kr/post/2021-04-26-mvc/) 글도 참고해보시면 좋을 것 같아요.

Rok93 commented on src/main/kotlin/week1/racingcar/RacingGame.kt:
```diff
@@ -2,35 +2,49 @@ package week1.racingcar
+        }
+    }
+
+    fun startGame() {
+        val cars = cars.map { Car(it) }
+        race(cars, timesToRace)
+    }
```
위에서 이야기한 것처럼 실제로 RacingGmae 객체의 주된 역할은 race 를 수행하는 것이라고 볼 수 있을 것 같은데, 
역할을 수행하기 위해서는 `List<Car>` 타입의 cars 가 필요한 것으로 보여요. 

실질직으로는 RacingGame이 역할을 수행하기위해 초기화하는 역할을 아직 수행하고있다고 볼 수 있어요. 👀

Rok93 replied:
위의 코멘트에 이어서 RacingGame의 역할을 수행하는 메서드는 무엇이라고 생각하시나요? 😃
아마도 그 메서드는 중요하기 때문에 테스트 코드를 특히나 꼭 작성해서 해야겠죠? 😉

Rok93 commented on src/main/kotlin/week1/racingcar/Car.kt:
```diff
@@ -1,10 +1,37 @@
+        }
+    }
+
+    fun moveOrNot() {
```
Car 객체의 public 메서드에 대해서 모두 테스트를 작성해보시면 좋을 것 같아요. 🙏

Rok93 reviewed (approved):
보성님 피드백 반영 잘해주셨습니다! 👍💯
이정도면 4단계는 충분한 것 같아서 이만 머지하겠습니다. 
특히 다음 미션과 관련된 코멘트들 남겨두었으니 참고하셔서 다음 단계에 고민해보시고 반영해보시면 좋을 것 같아요. 

또한 테스트 코드도 꼭 보충해주세요. 🙏
다음 단계에서 뵙겠습니다. 😉
미션 진행하시면서 어려운 부분이 있으면 언제든지 DM 주세요! 😃
